### PR TITLE
[Concurrency] Cleanup `async` signature and priority handling

### PR DESCRIPTION
**Pull request by DougGregor**
Fix a few related issues with `async` and priorities:
* Make `async` return a task handle. Overload on throwing/non-throwing operations for it to work out.
* Make the `priority` parameter to `async`, `asyncDetached`, `spawn`, and `spawnUnlessCancelled` optional, defaulting to `nil`.
* Deprecate `Task.Priority.unspecified`. This is better represented via optional `nil` in the APIs that use it.

### CODE REVIEW

**Comment by DougGregor** [comment]
@swift-ci please test

**Comment by DougGregor** [comment]
@swift-ci please test Linux

File: stdlib/public/Concurrency/Task.swift [hunk 1]
```diff
@@ -47,17 +47,21 @@ extension Task {
...
-      guard let task = task else {
-        return Priority.default
+      // If we are running on behalf of a task, use that task's priority.
+      if let task = task {
+        return task.priority
       }
 
-      return getJobFlags(task._task).priority
+      // Otherwise, query the system.
+      return Task.Priority(rawValue: _getCurrentThreadPriority()) ?? .default
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:64]: 👍

File: stdlib/public/Concurrency/Task.swift [hunk 2]
```diff
@@ -100,6 +104,8 @@ extension Task {
     case `default`       = 0x15
     case utility         = 0x11
     case background      = 0x09
+
+    @available(*, deprecated, message: "unspecified priority will be removed; use nil")
     case unspecified     = 0x00
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:109]: 👍

File: stdlib/public/Concurrency/Task.swift [hunk 3]
```diff
@@ -108,6 +114,18 @@ extension Task {
...
+extension Task.Priority {
+  /// Downgrade user-interactive to user-initiated.
+  var _downgradeUserInteractive: Task.Priority {
+    if self == .userInteractive {
+      return .userInitiated
+    }
+
+    return self
+  }
+}
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:127]: 👍

File: stdlib/public/Concurrency/Task.swift [hunk 4]
```diff
@@ -312,6 +330,22 @@ extension Task {
       }
     }
 
+    /// Whether this is a task created by the 'async' operation, which
+    /// conceptually continues the work of the synchronous code that invokes
+    /// it.
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:335]: oh okey, hm... i don't love the "continuing name" but it's better than no qualifier...

Will use this wording in the proposal too then 👍

File: stdlib/public/Concurrency/Task.swift [hunk 5]
```diff
@@ -438,23 +472,37 @@ public func detach<T>(
...
+  priority: Task.Priority,
+  @_inheritActorContext @_implicitSelfCapture operation: __owned @Sendable @escaping () async -> Void
+) {
+  let adjustedPriority: Task.Priority?
+  if priority == .unspecified {
+    adjustedPriority = nil
+  } else {
+    adjustedPriority = priority
+  }
+  let _: Task.Handle = async(priority: adjustedPriority, operation: operation)
```
Comment (ktoso) [inline stdlib/public/Concurrency/Task.swift:505]: Huh, without `<>`?

**Review by ktoso — APPROVED** [review]
LGTM, thanks for tackling this 👍

File: stdlib/public/Concurrency/Task.swift (hunk 5 above)
Comment (DougGregor) [reply stdlib/public/Concurrency/Task.swift:505]: Type inference is MAGIC

Comment (ktoso) [reply stdlib/public/Concurrency/Task.swift:505]: are you serious? LOL That's pretty both fun / hilarious / terrifying at the same time 🤣 

Interesting, did not know about that... Still tained by my scala days I guess, one had to write <_, _>

**Pull request merged**